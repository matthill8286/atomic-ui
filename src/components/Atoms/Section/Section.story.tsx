import { select } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { ThemeColors, VerticalPaddingMap } from '@/types/theme'
import { Section } from './Section'

const options = Object.keys(saiyanTheme.color) as ThemeColors[]

const knobs = (): { color: ThemeColors } => {
  return {
    color: select<ThemeColors>('Colors', options, 'primary'),
  }
}

const top: VerticalPaddingMap = {
  tablet: 'xs',
  desktop: 'md',
}

const bottom: VerticalPaddingMap = {
  mobile: 'lg',
  tablet: 'xl',
}

const height = {
  mobile: 350,
  tablet: 100,
  desktop: 400,
}

const bgImage = {
  mobile: '/public/images/featured_backgrounds/03.jpg',
  tablet: '/public/images/featured_backgrounds/01.jpg',
}

storiesOf('Design System/Atoms/Section', module)
  .add('Default', () => {
    return (
      <Section paddingTop={top} paddingBottom={bottom} {...knobs()}>
        Test
      </Section>
    )
  })
  .add('Fixed Height', () => {
    return (
      <Section paddingTop={top} paddingBottom={bottom} height={height} {...knobs()}>
        Test
      </Section>
    )
  })
  .add('with Image', () => {
    return (
      <Section paddingTop={top} image={bgImage} paddingBottom={bottom} height={height} {...knobs()}>
        Test
      </Section>
    )
  })
