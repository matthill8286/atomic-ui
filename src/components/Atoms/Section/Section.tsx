import { media } from '@/styles/media'
import { css, styled } from '@/styles/styled'
import { HeightMap, ImageMap, ThemeColors, VerticalPaddingMap } from '@/types/theme'
import { Shape, Corners } from '@/components/Atoms/Card/Card.interface'
import { handleShape } from '@/components/Atoms/Card/Card.styled'

export interface SectionProps {
  color?: ThemeColors
  image?: ImageMap | undefined
  paddingTop?: VerticalPaddingMap
  paddingBottom?: VerticalPaddingMap
  height?: HeightMap
  center?: boolean
  shape?: Shape | Corners<Shape>
}

const handlePadding = (paddingTop, paddingBottom) => {
  const { ...top } = paddingTop
  const { ...bottom } = paddingBottom

  return props => css`
    ${top.mobile && `padding-top: ${props.theme.spacing.base[top.mobile]}`};
    ${bottom.mobile && `padding-bottom: ${props.theme.spacing.base[bottom.mobile]}`};

    ${media.md} {
      ${top.tablet && `padding-top: ${props.theme.spacing.base[top.tablet]}`};
      ${bottom.tablet && `padding-bottom: ${props.theme.spacing.base[bottom.tablet]}`};
    }

    ${media.lg} {
      ${top.desktop && `padding-top: ${props.theme.spacing.base[top.desktop]}`};
      ${bottom.desktop && `padding-bottom: ${props.theme.spacing.base[bottom.desktop]}`};
    }
  `
}

export const Section = styled.section<SectionProps>`
  width: 100%;
  height: ${({ height }) => height && height.mobile && `${height.mobile}px`};
  left: 0;
  right: 0;
  background: ${({ color, theme }) => color && theme.color[color]} no-repeat center center;
  background-image: ${({ image }) => image && image.mobile && `url(${image.mobile}) `};
  background-size: cover;
  ${({ paddingTop, paddingBottom }) => handlePadding(paddingTop, paddingBottom)}
  ${({ center }) => center && `text-align: center;`}
  ${({ shape }) => shape && handleShape(shape)}

  ${media.md} {
    background-image: ${({ image }) => image && image.tablet && `url(${image.tablet}) `};
    height: ${({ height }) => height && height.tablet && `${height.tablet}px`};
  }

  ${media.lg} {
    background-image: ${({ image }) => image && image.desktop && `url(${image.desktop}) `};
    height: ${({ height }) => height && height.desktop && `${height.desktop}px`};
  }
`
