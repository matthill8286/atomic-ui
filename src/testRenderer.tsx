import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
