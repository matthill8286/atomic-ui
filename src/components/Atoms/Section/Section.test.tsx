import React from 'react'
import { renderWithTheme } from '@/testRenderer'
import { Section } from './Section'

describe('Divider', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(<Section color="primary" />)
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with paddings', () => {
    const tree = renderWithTheme(
      <Section
        color="primary"
        paddingTop={{ mobile: 'xs', tablet: 'lg' }}
        paddingBottom={{ mobile: 'lg', desktop: 'xs' }}
      />
    )
    expect(tree).toMatchSnapshot()
  })
})
