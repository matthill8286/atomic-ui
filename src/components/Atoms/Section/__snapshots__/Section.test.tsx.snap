// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider renders correctly 1`] = `
.c0 {
  width: 100%;
  left: 0;
  right: 0;
  background: #00C8DC no-repeat center center;
  background-size: cover;
}

@media (min-width:752px) {

}

@media (min-width:1008px) {

}

@media (min-width:752px) {

}

@media (min-width:1008px) {

}

<section
  className="c0"
  color="primary"
/>
`;

exports[`Divider renders correctly with paddings 1`] = `
.c0 {
  width: 100%;
  left: 0;
  right: 0;
  background: #00C8DC no-repeat center center;
  background-size: cover;
  padding-top: 8px;
  padding-bottom: 32px;
}

@media (min-width:752px) {
  .c0 {
    padding-top: 32px;
  }
}

@media (min-width:1008px) {
  .c0 {
    padding-bottom: 8px;
  }
}

@media (min-width:752px) {

}

@media (min-width:1008px) {

}

<section
  className="c0"
  color="primary"
/>
`;
